Identify and translate missing plural keys

markedmondson opened the pull request:
From https://github.com/glebm/i18n-tasks/issues/582

* For locales identified in a call to `translate` commands, determines the existing and required pluralization keys and adds the nodes to the forest before translating. It will additionally remove unrequired plural translation keys. 
* Adds a label to the front of missing plural keys when running the `missing` command as the table headers were a little unclear.
* Add a --skip-interpolation option to the `missing` command (but specifically intended for translation) that won't send **only** interpolated strings to translate, the reasoning behind this was a) I found OpenAI would end up getting a mismatched number of results returned if there were a lot of them and b) it's kind of wasteful as a correctly configured Rails implementation would fallback and work correctly.

I think there's still some cleanup to do here, it feels a little clobbered together as I was learning as I went so totally happy to solicit feedback (or hear and attempt to address any problems).

glebm commented on config/locales/en.yml:
```diff
@@ -23,6 +23,7 @@ en:
           nostdin: Do not read from stdin
           out_format: 'Output format: %{valid_text}'
           pattern_router: 'Use pattern router: keys moved per config data.write'
+          skip_interpolation: Skip translating strings that are straight interpolations (eg. "%{comment}")
```
Rather than adding an option, perhaps we should always simply copy such keys verbatim? There is never a reason to translate them.

markedmondson replied:
That would be preferential in our use case but I didn't want to assume...

markedmondson replied:
Actually we don't even copy them, we have Rails configured with fallback which then just works... Not sure that's a safe assumption either though.

glebm replied:
There is never any reason to translate interpolation-only values, so let's skip them unconditionally.

markedmondson replied:
Sounds good to me! :)

glebm replied:
Sure, makes sense to have that as an option.

glebm commented on lib/i18n/tasks/missing_keys.rb:
```diff
@@ -35,36 +35,48 @@ def eq_base_keys(opts = {})
-          present_keys = Set.new(children.map { |c| c.key.to_sym })
+
+          # Get the existing translated node if it exists
+          translated_node = data[locale].get("#{locale}." + node.full_key(root: false))
```
I think this is the same as:

```suggestion
          translated_node = data[locale].get(node.full_key(root: true))
```

markedmondson replied:
Let me double check but I think the goal here is to to copy the value from the locale we're iterating over from the passed locales in L60 (if it exists). If we used the node value, it'd give us the value from the base locale each time.

glebm replied:
Ah, that makes sense

markedmondson replied:
[f655f32](https://github.com/glebm/i18n-tasks/pull/596/commits/f655f32095f8c9c34c4327e769dfae67234527f4)

glebm commented on lib/i18n/tasks/translators/base_translator.rb:
```diff
@@ -102,6 +133,10 @@ def parse_value(untranslated, each_translated, opts)
           untranslated
         end
+      rescue StopIteration => e
+        require 'pry'
```
Left-over debugging code?

markedmondson replied:
Totally :)

Interestingly I was hitting it quite a bit using OpenAI when we had a batch of translations that was largely straight interpolations because it would return a mismatch of values, eg, we'd send 50 translation strings and get 49 back. Even specifically prompting to give the matching amount back, I guess ChatGPT is just bad at counting.

markedmondson commented on spec/translator_spec.rb:
```diff
@@ -0,0 +1,116 @@
+    end
+  end
+
+  describe 'multi-line' do
+
+  end
```
TODO: Remove

markedmondson commented on lib/i18n/tasks/translators/base_translator.rb:
```diff
@@ -51,7 +82,7 @@ def fetch_translations(list, opts)
       # @param [Array<[String, Object]>] list of key-value pairs
       # @return [Array<String>] values for translation extracted from list
       def to_values(list, opts)
-        list.map { |l| dump_value(l[1], opts) }.flatten.compact
+        list.map { |l| dump_value(l[1], opts) }.compact
```
This could probably go out as a "fix" on it's own - It was compacting array YAML values into the base array which caused the corresponding array size not to match (and also attempted to translate each line as an individual value)

markedmondson replied:
Seems like this might actually not be entirely correct.

Our usage is that array values are joined through Slim into paragraphs but obviously there can be cases where an array isn't intended to be used that way. 
We would expect the array to be compacted and translated in its whole context but there's an example in the test files of "encouragement" which doesn't seem to fit this model.

Thoughts on how to proceed?

markedmondson replied:
Got it :)

markedmondson commented on lib/i18n/tasks/missing_keys.rb:
```diff
@@ -93,9 +109,14 @@ def load_rails_i18n_pluralization!(locale)
+    def missing_diff_tree(locale, compared_to = base_locale, skip_interpolation = false)
+      data[compared_to].select_keys do |key, node|
+        # Skip if the string is just an interpolation, eg: "%{body}"
+        if skip_interpolation && node.value.to_s.match?(/^%{.*}$/)
```
TODO: Use `BaseTranslator::INTERPOLATION_KEY_RE`

markedmondson commented:
I spent a little more time on this on a large translation (~610 keys in 30 locales) and added a tweak.

Namely, ChatGPT is pretty bad at returning the correct number of responses back, I've added the array size to the prompt but even then it struggles if there are repeated translations - so there's a fail safe that'll retry and then fall back to doing line by line. I also found dropping the default array size from 50 to 25 improved things somewhat.

markedmondson commented:
One thing I noted which could perhaps be improved here is that it only adds missing plural keys, it doesn't remove. 

So what seems to happen is we add a new pluralization in en.yml with `one`, `other` and `zero` keys (`zero` being optional). i18n-tasks will translate those 2 (or 3) keys even when the corresponding locale might not need them. This then enforces the pluralization keys that are required (eg `one`, `few`, `many`, `other` in ru.yml) so we'll end up with `one`, `few`, `many`, `other`, and `zero` in ru.yml - the `zero` isn't used, however.

glebm commented on lib/i18n/tasks/base_task.rb:
```diff
@@ -39,6 +39,9 @@ class BaseTask
       include Stats
 
+      INTERPOLATION_KEY_RE = /%\{[^}]+}/.freeze
+      STRICT_INTERPOLATION_KEY_RE = /^%\{[^}]+\}$/.freeze
```
`INTERPOLATION_ONLY_KEY_RE`?

markedmondson replied:
[d15e17c](https://github.com/glebm/i18n-tasks/pull/596/commits/d15e17caa56dabfd7db6dcbe593e500de5b8c262)

markedmondson commented:
> Edit: Looks like there's a problem with the removal of pluralization keys that aren't required, eg we translate one, other and zero from en.yml, they get copied to ru.yml even though it only needs one, few, many, other because it gets caught and merged in the missing_diff_forest (even though missing_plural_forest removes it).

I fixed this in [7e47ded](https://github.com/glebm/i18n-tasks/pull/596/commits/7e47ded817d8909c7ea5464a29047eab86381d4a)
